apps/src/blockly/addons/plusMinusBlocks/if.js

```diff
@@ -0,0 +1,253 @@
+/**
+ * @license
+ * Copyright 2020 Google LLC
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * This file is based on code originally part of @blockly/block-plus-minus
+ * Modifications:
+ * - A minus field is added to else statements when created and on load
+ * - The plus field adds an else statement unless one already exists
+ */
+
+/**
+ * @fileoverview Changes the if block to use a +/- mutator UI.
+ */
+
+import GoogleBlockly from 'blockly/core';
+import {createMinusField} from './field_minus';
+import {createPlusField} from './field_plus';
+
+const controlsIfMutator = {
+  /**
+   * Number of else-if inputs on this block.
+   * @type {number}
+   */
+  elseIfCount_: 0,
+  /**
+   * Whether this block has an else input or not.
+   * @type {boolean}
+   */
+  hasElse_: false,
+
+  /**
+   * Creates XML to represent the number of else-if and else inputs.
+   * @returns {Element} XML storage element.
+   * @this {GoogleBlockly.Block}
+   */
+  mutationToDom: function () {
+    if (!this.elseIfCount_ && !this.hasElse_) {
+      return null;
+    }
+    const container = GoogleBlockly.utils.xml.createElement('mutation');
+    container.setAttribute('elseif', this.elseIfCount_);
+    if (this.hasElse_) {
+      // Has to be stored as an int for backwards compat.
+      container.setAttribute('else', 1);
+    }
+    return container;
+  },
+
+  /**
+   * Parses XML to restore the else-if and else inputs.
+   * @param {!Element} xmlElement XML storage element.
+   * @this {GoogleBlockly.Block}
+   */
+  domToMutation: function (xmlElement) {
+    const targetCount = parseInt(xmlElement.getAttribute('elseif'), 10) || 0;
+    this.hasElse_ = !!parseInt(xmlElement.getAttribute('else'), 10) || 0;
+    if (this.hasElse_ && !this.getInput('ELSE')) {
+      this.appendStatementInput('ELSE')
+        .appendField(GoogleBlockly.Msg['CONTROLS_IF_MSG_ELSE'])
+        .appendField(createMinusField(), 'MINUS_ELSE');
+    }
+    this.updateShape_(targetCount);
+  },
+
+  /**
+   * Returns the state of this block as a JSON serializable object.
+   * @returns {{elseIfCount: (number|undefined),
+   *     haseElse: (boolean|undefined)}} The state of this block, ie the else
+   *     if count and else state.
+   */
+  saveExtraState: function () {
+    if (!this.elseIfCount_ && !this.hasElse_) {
+      return null;
```

Comment:
should we return an empty object instead of `null`?

Reply:
Good question. This was the plugin’s return value and it does seem to work as is. I’ll investigate if there’s any difference as far as Blockly is concerned and consider what we do in other blocks as well.

Reply:
From https://developers.google.com/blockly/guides/create-custom-blocks/extensions#serialization_hooks:

> If your block is in its default state when it is serialized, then your `saveExtraState` method can return `null` to indicate this. If your `saveExtraState` method returns `null` then no `extraState` property is added to the JSON. This keeps your save file size small.

Going to leave this as is!